src/rez/tests/test_directives.py

```diff
@@ -0,0 +1,180 @@
+
+        # build the package from a clean build dir, then build it again
+        builder.build(clean=True)
+        builder.build()
+
+        # build and install it from a clean dir, then build and install it again
+        builder.build(install_path=self.install_root, install=True, clean=True)
+        builder.build(install_path=self.install_root, install=True)
+
+
+class TestBuildDirectives(_TestBuildDirectivesBase):
+
+    def test_build_soft(self):
+        _k = {"build_command": False}
+
+        self.dev_repo.add("dep", version="1.0.0", **_k)
+        self.dev_repo.add("dep", version="1.1.0", **_k)
+        self.dev_repo.add("var", version="2.1", **_k)
+        self.dev_repo.add("var", version="3.0", **_k)
+        self.dev_repo.add("lock", version="1", requires=["dep-1.0.0"], **_k)
+        self.dev_repo.add("soft",
+                          version="1",
+                          requires=["dep-1//harden(2)", "lock"],
+                          variants=[["var-2//harden(2)"], ["var-3.*"]],
+                          **_k)
+
+        self._test_build("dep", "1.0.0")
+        self._test_build("dep", "1.1.0")
+        self._test_build("var", "2.1")
+        self._test_build("var", "3.0")
+        self._test_build("lock", "1")
+        self._test_build("soft", "1")
+
+        soft = get_package("soft", "1", paths=[self.install_root])
+        self.assertEqual("dep-1.0", str(soft.requires[0]))
+
+        self.assertEqual(["var-2.1"], list(map(str, soft.variants[0])))
+        self.assertEqual(["var-3.0"], list(map(str, soft.variants[1])))
+
+    def test_build_soft_early(self):
+        _k = {"build_command": False}
+
+        self.dev_repo.add("dep", version="1.0.0", **_k)
+        self.dev_repo.add("dep", version="1.1.0", **_k)
+        self.dev_repo.add("lock", version="1", requires=["dep-1.0.0"], **_k)
+
+        @early()
+        def requires():
+            return (["dep-1//harden(2)", "lock"]
+                    if building else ["dep-1//harden(2)"])
+
+        self.dev_repo.add("soft", version="1", requires=requires, **_k)
+
+        self._test_build("dep", "1.0.0")
+        self._test_build("dep", "1.1.0")
+        self._test_build("lock", "1")
+        self._test_build("soft", "1")
+
+        soft = get_package("soft", "1", paths=[self.install_root])
+        self.assertEqual("dep-1.0", str(soft.requires[0]))
+
+    def test_empty_early_requires(self):
+        _k = {"build_command": False}
+
+        @early()
+        def requires():
+            return [] if building else ["bar-1//harden"]
+
+        self.dev_repo.add("bar", version="1.0.0", **_k)
+        self.dev_repo.add("foo", version="1.0.0", requires=requires, **_k)
+
+        self._test_build("bar", "1.0.0")
+        self._test_build("foo", "1.0.0")
+        foo = get_package("foo", "1.0.0", paths=[self.install_root])
+        # no harden, because there were no requires when building is True
+        self.assertEqual("bar-1", str(foo.requires[0]))
+
```

Comment:
Currently not sure how to deal with this, the early bound requires has a directive requirement but not appeared in build.
What I have in mind right now is to re-evaluate the package once more with `building=False`.

Reply:
And, harden requirement in late bound should be consider as missed used, correct ? wildcard require could use `expand_requirements` function, but since the harden is based on resolved packages, which means it will being harden anyway.